### PR TITLE
feat: [CDS-53513]: last published changes

### PR DESCRIPTION
**Pull request by shivamnegi94**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45880)

### CODE REVIEW

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java [hunk 1]
```diff
@@ -73,12 +80,38 @@ public ArtifactTaskExecutionResponse getLastSuccessfulBuild(BambooArtifactDelega
...
                                                             .build();
-    BuildDetails buildDetails = bambooBuildService.getLastSuccessfulBuild(null, artifactStreamAttributes,
-        BambooRequestResponseMapper.toBambooConfig(attributesRequest), attributesRequest.getEncryptedDataDetails());
-    BambooArtifactDelegateResponse bambooArtifactDelegateResponse =
-        BambooRequestResponseMapper.toBambooArtifactDelegateResponse(buildDetails, attributesRequest);
-    return getSuccessTaskExecutionResponse(
-        Collections.singletonList(bambooArtifactDelegateResponse), Collections.singletonList(buildDetails));
+
+    try {
+      Pattern pattern = Pattern.compile(attributesRequest.getBuildNumber());
```
Comment (puthrayaharness) [inline 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:85]: Why are we doing this?

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/custom/CustomArtifactService.java [hunk 2]
```diff
@@ -201,8 +204,10 @@ private ShellScriptTaskParametersNG getShellScriptTaskParametersNG(
 
   private List<BuildDetails> filterVersion(
       List<BuildDetails> buildDetails, CustomArtifactDelegateRequest attributesRequest) {
+    Pattern pattern = Pattern.compile(attributesRequest.getVersion());
     return buildDetails.stream()
-        .filter(build -> build.getNumber().equals(attributesRequest.getVersion()))
+        .filter(build -> pattern.matcher(build.getNumber()).find())
+        .sorted(new BuildDetailsComparatorDescending())
```
[the remark this answers is not in the document]
Comment (shivamnegi94) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/custom/CustomArtifactService.java:210]: To get the latest build on top

File: 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java (hunk 1 above)
Comment (shivamnegi94) [reply 930-delegate-tasks/src/main/java/io/harness/delegate/task/artifacts/bamboo/BambooArtifactTaskHandler.java:85]: In Bamboo we don't have option for build regex and we can pass regex as build number only.
Delegate side we are checking that if the build number is regex or not

**Comment by shivamnegi94** [comment]
trigger ut0, ut1, ut2, ut3, ut4

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

**Review by puthrayaharness — APPROVED** [review]

**Review by vtxorxwitty — APPROVED** [review]
> Reviewed 2 of 7 files at r1, 3 of 14 files at r2, 11 of 18 files at r3, 7 of 8 files at r4, 1 of 1 files at r5, 3 of 4 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45880)* status: 28 file(s) to review. 1. unresolved discussions.

**Pull request merged**